Add AfterShip sensor for packages

maxandersen opened the pull request:
Why:

  * I receive a lot of packages from many different shipping companies.
  * I would like to see in haas how many packages are being delivered.

  This change addreses the need by:

  * Adding a sensor for AfterShip (aftership.com)
  * AfterShip supports ~490 couriers world wide thus should cover
     almost any sensible tracking.

Notes:
  - For now this sensor assumes you somehow have added trackings to
      aftership manually
  - Future idea is to expose service that allows adding a tracking
        based on incoming mails.
  - Other improvments would be to add map markers for package locations.

 Related:
 - https://community.home-assistant.io/t/package-tracking/858
 - https://community.home-assistant.io/t/aftership-package-tracking/24068
 - https://community.home-assistant.io/t/aftership-shipment-tracking-platform/14074
 - https://community.home-assistant.io/t/aftership-state-card/57912

maxandersen committed:
Add AfterShip sensor for packages

Why:

 * I receive a lot of packages from many different shipping companies.
 * I would like to see in haas how many packages are being delivered.

This change addreses the need by:

 * Adding a sensor for AfterShip (aftership.com)
 * AfterShip supports ~490 couriers world wide thus should cover
   almost any sensible tracking.

Notes:
  - For now this sensor assumes you somehow have added trackings to
    aftership manually.
  - Future idea is to expose service that allows adding a tracking
    based on incoming mails.
  - Other improvments would be to add map markers for package locations.

Related:
- https://community.home-assistant.io/t/package-tracking/858
- https://community.home-assistant.io/t/aftership-package-tracking/24068
- https://community.home-assistant.io/t/aftership-shipment-tracking-platform/14074
- https://community.home-assistant.io/t/aftership-state-card/57912

maxandersen commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,117 @@
+
+    apikey = config.get(CONF_API_KEY)
+    name = config.get(CONF_NAME)
+    api = aftership.APIv4(apikey)
```
so do a call to the api, if fails what is the right way to make it fail ? just throw exception ?

and if works just throw the results out ?

maxandersen replied:
fixed by logging error and returning before adding devices/sensors.

maxandersen commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,117 @@
+sensor:
+   - platform: aftership
+     api_key: AFTERSHIP_API_KEY
+"""
```
will remove it.

fyi, I added this as the dev doc at https://developers.home-assistant.io/docs/en/development_checklist.html stated that was the way to start ?

maxandersen commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,117 @@
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the PostNL sensor platform."""
```
will fix! (on phone right now so will fix once back at laptop ;)

vMeph commented:
doesnt this sensor already exists
https://github.com/custom-cards/aftership-card

ludeeus commented:
@vMeph That is a custom card for Lovelace and not a sensor platfrom like this PR.
That card uses another custom_component to fetch the data since (currently) there is no dedicated sensor in Home Assistant for this.

vMeph commented:
ohh ok cool
ty for explanation

maxandersen commented:
yes, I learned after publishing this that https://github.com/custom-components/sensor.aftership exists and I think I'll merge its features. 

I do though think it would be nice to have a generic all-purpose package tracker (rather than the fairly limited/dedicated ones in there) - hence why suggesting to add it officially.

maxandersen commented:
one thing i'm wondering - is the best way to expose details about the current incoming packages really to put it as a json dictionary as data on the sensor or would it be better to have these as some kind of temporary tracking entities or similar ?

maxandersen commented:
fyi - delay in updating the PR is due to that after moving to hassio I'm battling this issue: https://community.home-assistant.io/t/runtimeerror-the-processor-time-used-is-not-available-or-its-value-cannot-be-represented/77659

alternatively I could just use the api directly (as its trivial) and avoid the use of clock() in the aftership sdk - suggestions welcome.

frenck commented:
Could not find a related PR on our documentation repository. Adding `docs-missing` label.

maxandersen commented:
Now cleaned this up based on feedback.

I also had to move to using pyaftership rather than aftership's own python sdk as the latter fails when running over a day on a hass.io raspberry (some issue using .clock() calls).

In addition I added a service to call for adding packages - I wanted to add service description but did not find services.yaml for sensors - is that expected ? should I create a new ? 

I'll make a corresponding doc tomorrow - time to sleep now ;)

fabaff commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,161 @@
+
+TITLE = 'title'
+SLUG = 'slug'
+TRACKING_NUMBER = 'tracking_number'
```
Please prefix then with `CONF_`.

maxandersen replied:
done.

fabaff commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,161 @@
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+_LOGGER = logging.getLogger(__name__)
```
Already present on line 32.

maxandersen replied:
fixed.

fabaff commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,161 @@
+from homeassistant.util import Throttle
+from homeassistant.exceptions import PlatformNotReady
+
+REQUIREMENTS = ['pyaftership==0.0.5']
```
It looks like that the latest release is `0.1.1`.

maxandersen replied:
yeah, new update came out based on this PR - pushed update using latest that is also async. 

but just noticed the service is failing for this version so looking into that now.

maxandersen replied:
removed the service code to make the initial PR simpler.

maxandersen commented:
apparently my githook isn't properly running  lint ;/ fix those issues and pushed updated that has async and latest pyaftership usage.

maxandersen commented:
Just noticed the 17track component which does the sensor per package I wanted to do. 

I know you prefer small PRS over big ones. Could we meet the current state for this one and then I add the package sensors or you prefer I do it in one go ? Thanks!

MartinHjelmare commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,130 @@
+                      aftership.meta)
+        return
+
+    sensor = [AfterShipSensor(aftership, name)]
```
This is a list so perhaps call it sensors, or move the list wrapping to the add entities call.

maxandersen replied:
fixed.

MartinHjelmare commented on homeassistant/components/sensor/aftership.py:
```diff
@@ -0,0 +1,130 @@
+        for tracking in self.aftership.trackings['trackings']:
+            status = tracking['tag']
+            name = tracking['tracking_number']
+            status_counts[status] = status_counts.get(status, 0)+1
```
Is the status key lowercase snakecase? That's our standard for state attribute keys.

maxandersen replied:
the values are camel case with spaces in aftership api, see list at https://help.aftership.com/hc/en-us/articles/360007823253

want to lower case it here when storing in state ?

MartinHjelmare replied:
That would be good!

balloob replied:
Please update this and I think this PR is ready to merge.

maxandersen replied:
sorry missed the ping - pushed an update that fixes this. it just stores lower values now.

MartinHjelmare commented:
This is almost ready for merge. Please don't extend it more in this PR, unless it would be a breaking change later.

maxandersen commented:
I'm cool with leave as is if that gets it included; then I can add the actual individual package tracking in additional PR's.

MartinHjelmare reviewed (approved):
Good!

For the future, please don't squash commits after review has started to make it easier for readers to track changes. We always squash upon merge via github anyway.

Can be merged when build passes.

fabaff committed:
Fix typo and update ordering

MartinHjelmare commented:
Thanks @fabaff!
